[cache_key_fields] panic: interface conversion: interface {} is nil, not map[string]interface

### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

Terraform: 1.3.5
Cloudflare plugin: 3.28.0 

### Affected resource(s)

cloudflare_page_rule -> cache_key_fields

### Terraform configuration files

```hcl
resource "cloudflare_page_rule" "page_rule_4086xxxf46f9" {
    zone_id = var.zone_id

    target = "www.example.com/*"

    priority = 6
    status = "active"
    actions {
    
        cache_level = "cache_everything"
    
        edge_cache_ttl = 18000
    
        cache_key_fields {
            query_string {
                exclude = ["a_xxx","b_xxx","c_tid","d_xxxx","e_xxx","f_xxx","fbclid","xid","xhash","msclkid","dclid","gclid"]
            }
            header {
                include = ["authorization"]
            }
            cookie {}
            user {}
            host {}
        }
    }
    depends_on = [cloudflare_page_rule.page_rule_d82xxxxx6c5f4e1]
}
```


### Link to debug output

https://gist.github.com/Kevin-Li-Webjet/50f85989088e63464f7ef176072fcd08

### Panic output

Stack trace from the terraform-provider-cloudflare_v3.28.0 plugin:

panic: interface conversion: interface {} is nil, not map[string]interface {}

goroutine 43 [running]:
github.com/cloudflare/terraform-provider-cloudflare/internal/provider.transformToCloudflarePageRuleAction({0x10360c8, 0xc0005248a0}, {0xebcf3a, 0x10}, {0xd46900?, 0xc0005f2048?}, 0x8?)
        github.com/cloudflare/terraform-provider-cloudflare/internal/provider/resource_cloudflare_page_rule.go:466 +0x11f7
github.com/cloudflare/terraform-provider-cloudflare/internal/provider.resourceCloudflarePageRuleUpdate({0x10360c8, 0xc0005248a0}, 0xc000582780, {0xeae680?, 0xc0000c03c0?})
        github.com/cloudflare/terraform-provider-cloudflare/internal/provider/resource_cloudflare_page_rule.go:176 +0x6fa
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc00016eb60, {0x1036100, 0xc0004b5a70}, 0xd?, {0xeae680, 0xc0000c03c0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:741 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00016eb60, {0x1036100, 0xc0004b5a70}, 0xc000393040, 0xc0000ad200, {0xeae680, 0xc0000c03c0})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:847 +0x83a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000010078, {0x1036100?, 0xc0004b51d0?}, 0xc0000b0460)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000256a00, {0x1036100?, 0xc000186630?}, 0xc000258310)
        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xe7dfe0?, 0xc000256a00}, {0x1036100, 0xc000186630}, 0xc000258070, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0004ae000, {0x1039d00, 0xc0004f8680}, 0xc00033e000, 0xc0003038c0, 0x1599120, 0x0)
        google.golang.org/grpc@v1.50.1/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0004ae000, {0x1039d00, 0xc0004f8680}, 0xc00033e000, 0x0)
        google.golang.org/grpc@v1.50.1/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.50.1/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.50.1/server.go:963 +0x28a

Error: The terraform-provider-cloudflare_v3.28.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

### Expected output

changes applied

### Actual output

 Error: Plugin did not respond
│ 
│   with module.xxxx.cloudflare_page_rule.page_rule_63896b0561214cbeb2ec384bf2ef36a0,
│   on accounts/xxxx/xxxx/page_rules.tf line 229, in resource "cloudflare_page_rule" "page_rule_63896b056xxxx4bf2ef36a0":
│  229: resource "cloudflare_page_rule" "page_rule_63896b056xxxxx4bf2ef36a0" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs
│ may contain more details.

### Steps to reproduce

take my config code
terraform init
terraform plan
terraform apply

### Additional factoids

_No response_

### References

similar with https://github.com/hashicorp/terraform-provider-aws/issues/24881

2022-11-29T22:37:25.110Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: cache_key_fields action to be applied: []interface {}{map[string]interface {}{"cookie":[]interface {}{map[string]interface {}{"check_presence":*Set(map[string]interface {}(nil)), "include":*Set(map[string]interface {}(nil))}}, "header":[]interface {}{map[string]interface {}{"check_presence":*Set(map[string]interface {}(nil)), "exclude":*Set(map[string]interface {}(nil)), "include":*Set(map[string]interface {}{"1405523147":"authorization"})}}, "host":[]interface {}{map[string]interface {}{"resolved":false}}, "query_string":[]interface {}{map[string]interface {}{"exclude":*Set(map[string]interface {}{"1427741591":"utm_term", "1758001918":"fbclid", "2018762823":"xid", "2052660961":"dclid", "2220716248":"utm_tid", "223125000":"utm_source", "2447139227":"utm_medium", "2850690849":"utm_content", "4241321039":"gclid", "566336536":"utm_campaign", "921918567":"xhash", "978681895":"msclkid"}), "ignore":false, "include":*Set(map[string]interface {}(nil))}}, "user":[]interface {}{interface {}(nil)}}}: timestamp=2022-11-29T22:37:25.110Z
2022-11-29T22:37:25.115Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: panic: interface conversion: interface {} is nil, not map[string]interface {}
2022-11-29T22:37:25.115Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: 
2022-11-29T22:37:25.115Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: goroutine 27 [running]:
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/cloudflare/terraform-provider-cloudflare/internal/provider.transformToCloudflarePageRuleAction({0x10360c8, 0xc0005576e0}, {0xebcf3a, 0x10}, {0xd46900?, 0xc0002eef48?}, 0x8?)
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/cloudflare/terraform-provider-cloudflare/internal/provider/resource_cloudflare_page_rule.go:466 +0x11f7
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/cloudflare/terraform-provider-cloudflare/internal/provider.resourceCloudflarePageRuleUpdate({0x10360c8, 0xc0005576e0}, 0xc000625300, {0xeae680?, 0xc000126600?})
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/cloudflare/terraform-provider-cloudflare/internal/provider/resource_cloudflare_page_rule.go:176 +0x6fa
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc00018aa80, {0x1036100, 0xc000760d50}, 0xd?, {0xeae680, 0xc000126600})
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:741 +0x12e
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00018aa80, {0x1036100, 0xc000760d50}, 0xc0002bac30, 0xc000601600, {0xeae680, 0xc000126600})
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:847 +0x83a
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000462000, {0x1036100?, 0xc000760c30?}, 0xc000587360)
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/grpc_provider.go:1021 +0xe8d
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000000b40, {0x1036100?, 0xc000259890?}, 0xc000606a10)
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/tf5server/server.go:818 +0x574
2022-11-29T22:37:25.116Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0xe7dfe0?, 0xc000000b40}, {0x1036100, 0xc000259890}, 0xc0006069a0, 0x0)
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000476000, {0x1039d00, 0xc0003189c0}, 0xc0001c6240, 0xc0003019b0, 0x1599120, 0x0)
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        google.golang.org/grpc@v1.50.1/server.go:1340 +0xd23
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: google.golang.org/grpc.(*Server).handleStream(0xc000476000, {0x1039d00, 0xc0003189c0}, 0xc0001c6240, 0x0)
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        google.golang.org/grpc@v1.50.1/server.go:1713 +0xa2f
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: google.golang.org/grpc.(*Server).serveStreams.func1.2()
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        google.golang.org/grpc@v1.50.1/server.go:965 +0x98
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0: created by google.golang.org/grpc.(*Server).serveStreams.func1
2022-11-29T22:37:25.117Z [DEBUG] provider.terraform-provider-cloudflare_v3.28.0:        google.golang.org/grpc@v1.50.1/server.go:963 +0x28a
2022-11-29T22:37:25.126Z [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2022-11-29T22:37:25.126Z [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/cloudflare/cloudflare/3.28.0/linux_amd64/terraform-provider-cloudflare_v3.28.0 pid=30 error="exit status 2"
2022-11-29T22:37:25.127Z [ERROR] plugin.(*GRPCProvider).ApplyResourceChange: error="rpc error: code = Unavailable desc = error reading from server: EOF"
2022-11-29T22:37:25.139Z [ERROR] vertex "module.xxx.cloudflare_page_rule.page_rule_63896b0561xxxxf2ef36a0" error: Plugin did not respond
╷
│ Error: Plugin did not respond
│ 
│   with module.xxx.cloudflare_page_rule.page_rule_63896b056xxxxb2ec384bf2ef36a0,
│   on accounts/xxx/xxx/page_rules.tf line 229, in resource "cloudflare_page_rule" "page_rule_63896b0561214cbeb2ec384bf2ef36a0":
│  229: resource "cloudflare_page_rule" "page_rule_63896b056121xxxxx384bf2ef36a0" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
2022-11-29T22:37:25.460Z [DEBUG] Azure Backend Request: 
HEAD /cloudflare/terraform.tfstate HTTP/1.1
Host: wjtfstates.blob.core.windows.net
User-Agent: HashiCorp Terraform/1.3.5 (+https://www.terraform.io)
X-Ms-Date: Tue, 29 Nov 2022 22:37:25 GMT
X-Ms-Lease-Id: 45f069c1-e955-c4fb-d9f9-a6b82171d6f8
X-Ms-Version: 2018-11-09
2022-11-29T22:37:25.480Z [DEBUG] Azure Backend Response for https://xxx.blob.core.windows.net/cloudflare/terraform.tfstate: 
HTTP/1.1 200 OK
Content-Length: 1300768

--- a/.changelog/2208.txt
+++ b/.changelog/2208.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/cloudflare_page_rule: remove empty cookie and header fields when applying this resource
+```

--- a/internal/sdkv2provider/resource_cloudflare_page_rule.go
+++ b/internal/sdkv2provider/resource_cloudflare_page_rule.go
@@ -290,8 +290,24 @@ func transformFromCloudflarePageRuleAction(pageRuleAction *cloudflare.PageRuleAc
 
 		for sectionID, sectionValue := range pageRuleAction.Value.(map[string]interface{}) {
 			switch sectionID {
-			case "cookie", "header", "host", "user":
+			case "host", "user":
 				output[sectionID] = []interface{}{sectionValue}
+
+			case "cookie", "header":
+				fieldOutput := map[string]interface{}{}
+				for fieldID, fieldValue := range sectionValue.(map[string]interface{}) {
+					switch fieldValue.(type) {
+					case []interface{}:
+						if len(fieldValue.([]interface{})) > 0 {
+							fieldOutput[fieldID] = fieldValue
+						}
+					default:
+						fieldOutput[fieldID] = fieldValue
+					}
+				}
+				if len(fieldOutput) > 0 {
+					output[sectionID] = []interface{}{fieldOutput}
+				}
 
 			case "query_string":
 				fieldOutput := map[string]interface{}{}
@@ -422,11 +438,14 @@ func transformToCloudflarePageRuleAction(ctx context.Context, id string, value i
 
 				switch sectionID {
 				case "cookie", "header":
-					for fieldID, fieldValue := range sectionValue.([]interface{})[0].(map[string]interface{}) {
-						sectionOutput[fieldID] = fieldValue.(*schema.Set).List()
+					if len(sectionValue.([]interface{})) > 0 && sectionValue.([]interface{})[0] != nil {
+						for fieldID, fieldValue := range sectionValue.([]interface{})[0].(map[string]interface{}) {
+							sectionOutput[fieldID] = fieldValue.(*schema.Set).List()
+						}
+						output[sectionID] = sectionOutput
 					}
 				case "query_string":
-					if sectionValue.([]interface{})[0] != nil {
+					if len(sectionValue.([]interface{})) > 0 && sectionValue.([]interface{})[0] != nil {
 						for fieldID, fieldValue := range sectionValue.([]interface{})[0].(map[string]interface{}) {
 							switch fieldID {
 							case "exclude", "include":
@@ -464,12 +483,13 @@ func transformToCloudflarePageRuleAction(ctx context.Context, id string, value i
 
 					output[sectionID] = sectionOutput
 				default:
-					for fieldID, fieldValue := range sectionValue.([]interface{})[0].(map[string]interface{}) {
-						sectionOutput[fieldID] = fieldValue
+					if len(sectionValue.([]interface{})) > 0 {
+						for fieldID, fieldValue := range sectionValue.([]interface{})[0].(map[string]interface{}) {
+							sectionOutput[fieldID] = fieldValue
+						}
+						output[sectionID] = sectionOutput
 					}
 				}
-
-				output[sectionID] = sectionOutput
 			}
 
 			pageRuleAction.Value = output

--- a/internal/sdkv2provider/resource_cloudflare_page_rule_test.go
+++ b/internal/sdkv2provider/resource_cloudflare_page_rule_test.go
@@ -738,6 +738,39 @@ func TestAccCloudflarePageRuleCacheKeyFieldsIncludeMultipleValuesQueryString(t *
 	})
 }
 
+func TestAccCloudflarePageRule_EmptyCookie(t *testing.T) {
+	var pageRule cloudflare.PageRule
+	domain := os.Getenv("CLOUDFLARE_DOMAIN")
+	zoneID := os.Getenv("CLOUDFLARE_ZONE_ID")
+	rnd := generateRandomResourceName()
+	pageRuleTarget := fmt.Sprintf("%s.%s", rnd, domain)
+	resourceName := fmt.Sprintf("cloudflare_page_rule.%s", rnd)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: providerFactories,
+		CheckDestroy:      testAccCheckCloudflarePageRuleDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflarePageRuleEmtpyCookie(zoneID, rnd, pageRuleTarget),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudflarePageRuleExists(resourceName, &pageRule),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.host.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.query_string.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.user.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.cookie.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.header.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.host.0.resolved", "true"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.query_string.0.include.#", "0"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.user.0.device_type", "true"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.user.0.geo", "false"),
+					resource.TestCheckResourceAttr(resourceName, "actions.0.cache_key_fields.0.user.0.lang", "false"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccCloudflarePageRuleCacheTTLByStatus(t *testing.T) {
 	var pageRule cloudflare.PageRule
 	domain := os.Getenv("CLOUDFLARE_DOMAIN")
@@ -1283,4 +1316,27 @@ func testAccCheckCloudflarePageRuleHasAction(pageRule *cloudflare.PageRule, key 
 		}
 		return fmt.Errorf("cloudflare page rule action not found %#v:%#v\nAction State\n%#v", key, value, pageRule.Actions)
 	}
+}
+
+func testAccCheckCloudflarePageRuleEmtpyCookie(zoneID, rnd, target string) string {
+	return fmt.Sprintf(`
+resource "cloudflare_page_rule" "%[3]s" {
+	zone_id = "%[1]s"
+	target = "%[3]s"
+	actions {
+    cache_key_fields {
+      host {
+        resolved = true
+      }
+      query_string {
+        ignore = true
+      }
+      user {
+        device_type = true
+        geo         = false
+        lang        = false
+      }
+    }
+  }
+}`, zoneID, target, rnd)
 }

--- a/internal/sdkv2provider/schema_cloudflare_page_rule.go
+++ b/internal/sdkv2provider/schema_cloudflare_page_rule.go
@@ -350,7 +350,7 @@ func resourceCloudflarePageRuleSchema() map[string]*schema.Schema {
 
 								"host": {
 									Type:     schema.TypeList,
-									Optional: true,
+									Required: true,
 									MinItems: 1,
 									MaxItems: 1,
 									Elem: &schema.Resource{
@@ -366,7 +366,7 @@ func resourceCloudflarePageRuleSchema() map[string]*schema.Schema {
 
 								"query_string": {
 									Type:     schema.TypeList,
-									Optional: true,
+									Required: true,
 									MinItems: 1,
 									MaxItems: 1,
 									Elem: &schema.Resource{
@@ -398,7 +398,7 @@ func resourceCloudflarePageRuleSchema() map[string]*schema.Schema {
 
 								"user": {
 									Type:     schema.TypeList,
-									Optional: true,
+									Required: true,
 									MinItems: 1,
 									MaxItems: 1,
 									Elem: &schema.Resource{

